volmdlr/curves.py

```diff
@@ -2371,17 +2371,179 @@ def point_over_contour(self, point, abs_tol=1e-6):
         :return: The distance between ellipse and point
         :rtype: float.
         """
-        start = self.point_at_abscissa(0.0)
-        return vm_common_operations.get_point_distance_to_edge(self, point, start, start)
+        point_local = self.frame.global_to_local_coordinates(point)
+        a = self.major_axis
+        b = self.minor_axis
+        x = point_local.x
+        y = point_local.y
+
+        if x >= 0:
+            initial_angle = 0
+        else:
+            initial_angle = math.pi
+
+        if y == 0 and abs(x) < (a * a - b * b) / a:
+            if x == 0:
+                return b
+            abscissa = math.acos((a * x) / (a * a - b * b))
+            ellipse_point0 = self.point_at_polar_abscissa(abscissa)
+            ellipse_point1 = self.point_at_polar_abscissa(initial_angle)
+            if ellipse_point0.point_distance(point) < ellipse_point1.point_distance(point):
+                return ellipse_point0.point_distance(point)
+            return ellipse_point1.point_distance(point)
+
+        ellipse_point_ = self.nearest_point(point)
+        return ellipse_point_.point_distance(point)
+
+    def tangent_points(self, point):
+        """
+        Calculates the points witch form a tangent line between the ellipse and an other point.
+
+        :param point: Other point to calculate distance.
+        :type point: volmdlr.Point2D.
+        :return: Two points on ellipse
+        :rtype: [Point2D, Point2D].
+        """
+        if self.point_inside(point) or self.point_belongs(point):
+            return [point, point]
+        a = self.major_axis
+        b = self.minor_axis
+
+        local_nearest_point = self.frame.global_to_local_coordinates(self.nearest_point(point))
+        u1, u2 = local_nearest_point.y / b, local_nearest_point.x / a
+        nearest_angle = math.atan2(u1, u2)
+        contour_angle_1 = nearest_angle + math.pi
+        contour_angle_2 = nearest_angle - math.pi
+        midle_angle = nearest_angle
+        j = 0
+        # bisection method
+        while j < 100:
+            new_angle = (midle_angle + contour_angle_1) / 2
+            product = self.vectorial_product(point, new_angle)
+            if product > 0:
+                contour_angle_1 = new_angle
+            else:
+                midle_angle = new_angle
+            if abs(product) <= 1e-7:
+                break
+            j += 1
+        point_1 = self.point_at_polar_abscissa(new_angle)
+
+        midle_angle = nearest_angle
+        j = 0
+        # bisection method
+        while j < 100:
+            new_angle = (midle_angle + contour_angle_2) / 2
+            product = self.vectorial_product(point, new_angle)
+            if product > 0:
+                contour_angle_2 = new_angle
+            else:
+                midle_angle = new_angle
+            if abs(product) <= 1e-7:
+                break
+            j += 1
+        point_2 = self.point_at_polar_abscissa(new_angle)
```

Comment:
could you simplify this code (also repeated above) in a separate method? like `def bisection_point(middle_angle, contour_angle)` or something like that, which will return the point? it is just to avoid repeating code